rpc, gui: bumpfee signer support

Sjors opened the pull request:
The `bumpfee` RPC call and GUI fee bump interface now work with an external signer.

Sjors commented:
I'm not tremendously excited by this approach of overloading `bumpfee`. But it also seems wrong to use `psbtbumpfee` here for the simple case of a single signer than has all the required keys. I could add yet another RPC method `signerbumpfee`?

cc @achow101 @instagibbs

meshcollider reviewed (commented):
utACK 0c0d293d31e96c9119e4793df66b375ec9f8495e

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3521,16 +3521,39 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
     // If wallet private keys are enabled, return the new transaction id,
     // otherwise return the base64-encoded unsigned PSBT of the new transaction.
```
Do the merged changes in #22021 need to be updated with this pull?

Sjors replied:
I'll do a rebase and also address the nits above.

Sjors replied:
For now these comments are correct. In a multisig setup `bumpfee` will throw ` "Transaction incomplete. Try psbtbumpfee instead."`. We can improve that later.

Sjors committed:
rpc: document bools in FillPSBT() calls

Sjors commented:
Rebased, addressed comments and added a test.

meshcollider reviewed (commented):
re-utACK dda02ce4f99cff37bbd41373df6ecb4541ed66f5

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4174,10 +4195,10 @@ static RPCHelpMan send()
             bool complete;
-            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, false, true);
-            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, true, false);
+            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, false /* sign */, true /* bip32derivs */);
```
Usually we put the meaning-comment before the value.

Sjors replied:
That indeed seems more common (`*/,` has 93 results vs 441 for `, /* `)

luke-jr commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3530,16 +3530,37 @@ static RPCHelpMan bumpfee_helper(std::string method_name)
+                throw JSONRPCError(RPC_WALLET_ERROR, "Transaction incomplete. Try psbtbumpfee instead.");
+            }
+            CTransactionRef tx(MakeTransactionRef(std::move(mtx)));
+            result.pushKV("txid", tx->GetHash().GetHex());
```
Prefer to assign `txid` here and do a single shared `pushKV` later.

luke-jr reviewed (approved):
utACK

luke-jr commented:
On further reflection, any reason not to put these details in `feebumper::SignTransaction` so it's more easily reused in the GUI?

Sjors commented:
@luke-jr good idea: now it works with GUI too!

Sjors commented:
Rebased after #23667 (manually re-applied changes from `wallet/rpcwallet.cpp` to `rpc/spend.cpp`)

achow101 commented:
There was a bad rebase, you've re-added `src/wallet/rpcwallet.cpp`.

Sjors commented:
Oops, fixed.

achow101 commented:
ACK d5f720fb48d3eeb64101af1d5614a3650643bb1e

Sjors commented:
Rebased after #441, though not tested yet.

luke-jr commented on src/wallet/rpc/spend.cpp:
```diff
@@ -1121,10 +1124,10 @@ RPCHelpMan send()
-            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, false, true);
-            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, true, false);
+            pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, /* sign */ false, /* bip32derivs */ true);
+            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, /* sign */ true, /* bip32derivs */ false);
```
```suggestion
            const TransactionError err = pwallet->FillPSBT(psbtx, complete, SIGHASH_DEFAULT, /*sign=*/true, /*bip32derivs=*/false);
```

& other such changes

Sjors replied:
Fixed (after the rebase moved everything around)

Sjors commented:
Ok, it works again now.

Sjors committed:
rpc: bumpfee signer support

Sjors committed:
gui: bumpfee signer support

Specifically this enables the Send button in the fee bump dialog for wallets with external signer support. Similar to 2efdfb88aab6496dcf2b98e0de30635bc6bade85.

furszy commented on src/wallet/feebumper.cpp:
```diff
@@ -239,7 +239,22 @@ Result CreateRateBumpTransaction(CWallet& wallet, const uint256& txid, const CCo
-    return wallet.SignTransaction(mtx);
+
+    if (wallet.IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
+        // Make a blank psbt
+        PartiallySignedTransaction psbtx(mtx);
+
+        // First fill transaction with our data without signing,
+        // so external signers are not asked to sign more than once.
+        bool complete;
+        wallet.FillPSBT(psbtx, complete, SIGHASH_ALL, false /* sign */, true /* bip32derivs */);
+        const TransactionError err = wallet.FillPSBT(psbtx, complete, SIGHASH_ALL, true /* sign */, false  /* bip32derivs */);
+        if (err != TransactionError::OK) return false;
+        complete = FinalizeAndExtractPSBT(psbtx, mtx);
+        return complete;
```
nit: as this is the same code that we have in "rpc/send" --> `FinishTransaction`, could decouple it into a new function and use it from both sides.

furszy reviewed (commented):
code review ACK 2c07cfac

jarolrod reviewed (commented):
Approach ACK

I was able to successfully sign for a fee increase on my external signer. Want to continue to do more testing.

achow101 commented:
ACK 2c07cfacd1745844a1d3c57f2e8617549b9815d7

jarolrod reviewed (approved):
tACK 2c07cfa

I have tested and reviewed the code and agree it can be merged.
